frb_example/pure_dart/dart/lib/bridge_generated.dart

```diff
@@ -346,6 +382,36 @@ class ConcatenateWith {
       bridge.handleSomeStaticStreamSinkSingleArgStaticMethodConcatenateWith(hint: hint);
 }
 
+@freezed
+class CustomError with _$CustomError {
+  const factory CustomError.error0(
```

Comment:
In addition, IMHO stack trace is very important data and helps a lot in debugging. Shall we do something about it?

For example, https://docs.rs/thiserror/latest/thiserror/ does seem to provide a `backtrace`

Reply:
I added support for backtrace on Rust and Dart. On Dart I did:

class Backtrace {
  Backtrace(this.backtrace);
  String backtrace;
  @override
  String toString() => backtrace;
}

```dart
class FrbException {
  Backtrace? backtrace;

  void setBacktrace(Backtrace backtrace) {
    this.backtrace = backtrace;
  }

  Backtrace? getBacktrace() {
    return backtrace;
  }

  @override
  String toString() => 'FrbException, backtrace $backtrace';
}
```

then

```dart
@freezed
class CustomError extends FrbException with _$CustomError implements Exception {
  const factory CustomError.error0(
    String field0,
  ) = Error0;
  const factory CustomError.error1(
    int field0,
  ) = Error1;
}
```

but on

`class _$Error0 implements Error0 {`

the `_$Error0` gives an error:

```
lib/bridge_generated.freezed.dart:667:7: Error: The non-abstract class '_$Error1' is missing implementations for these members:
 - FrbException.backtrace
 - FrbException.setBacktrace
Try to either
 - provide an implementation,
 - inherit an implementation from a superclass or mixin,
 - mark the class as abstract, or
 - provide a 'noSuchMethod' implementation.
```

I don't know what to do here. Can't do any of the suggestions. This is the only thing missing now.

Reply:
>  extends FrbException with _$CustomError implements Exception

Oops if `extends FrbException`, we only need to let `class FrbException implements Exception` and we are done. No need for `implements Exception` here anymore.

Reply:
Btw 

```
class FrbException {
  Backtrace? backtrace;

  void setBacktrace(Backtrace backtrace) {
    this.backtrace = backtrace;
  }

  Backtrace? getBacktrace() {
    return backtrace;
  }

  @override
  String toString() => 'FrbException, backtrace $backtrace';
}
```

is not very Dart-taste. What about something like:

```
class Backtrace {
  Backtrace(this.backtrace);
  final String backtrace; // <--- NOTE be final
  @override
  String toString() => backtrace;
}

class FrbException {
  final Backtrace? backtrace; // <--- NOTE be final

  FrbException({required this.backtrace});

  @override
  String toString() => 'FrbException(backtrace: $backtrace)';
}
```

Reply:
```dart
class FrbException implements Exception {
  Backtrace? backtrace;

  FrbException();
  //FrbException({required this.backtrace}); // Tried this but then `CustomError` needs to explicitly set a backtrace, but a backtrace should be optional

  void setBacktrace(Backtrace backtrace) {
    this.backtrace = backtrace;
  }

  Backtrace? getBacktrace() {
    return backtrace;
  }

  @override
  String toString() => 'FrbException(backtrace: $backtrace)';
}
```

```dart
@freezed
class CustomError extends FrbException with _$CustomError {
  const factory CustomError.error0(
    String field0,
  ) = Error0;
  const factory CustomError.error1(
    int field0,
  ) = Error1;
}
```

then

`abstract class Error0 implements CustomError {`

then

`class _$Error0 implements Error0 {`

but `_$Error0` still gives 

```
Missing concrete implementation of 'getter FrbException.backtrace'.
Try implementing the missing method, or make the class abstract.
```

maybe because the class is abstract?

Reply:
Oops, then what about this:

1. `FrbException` has *no fields*. It is almost blank, and purely serve as base class for all exceptions that our frb may throw.
2. `abstract class FrbBacktracedException implements FrbException {final Backtrace backtrace;}` is a frb exception with backtrace
3. For anyhow error, maybe be something like `class FrbAnyhowException extends FrbBacktraceException` and we can provide backtrace. Another solution may be treat anyhow error without special care.
4. For custom errors, just `implements FrbException` (which should have no problem since it is blank).
5. For custom errors *that does have a backtrace field*, maybe use the `@Implements<FrbBacktracedException>()` mentioned in https://pub.dev/packages/freezed

Reply:
but things like `class _$Error0 implements Error0 {` are generated automatically, adding `@Implements<FrbBacktracedException>()` is not possible I imagine. 

Also currently a anyhow is just a String on the Dart side, `-> Result<(), anyhow::Error>` gets converted to `Box<dyn FnOnce()->DartCObject>`. It all depends on what is chosen for the output of anyhow, which is a simple `.to_string().into_dart()` right know

Reply:
> adding @Implements<FrbBacktracedException>() is not possible I imagine.

Have a look at freezed doc - it is annotated on `const factory CustomError.error0({})`

> Also currently a anyhow is just a String on the Dart side

Sure. Shall we change something?

Reply:
I made FrbException abstract and having nothing, but I'm confused.

"Note 2: You cannot use @With/@Implements with freezed classes. Freezed classes can neither be extended nor implemented."

Also, I'm very confused, I'm not good at Dart. Do you mean something like this?

```
@freezed
class CustomNestedError1 with _$CustomNestedError1 implements FrbException {
  @Implements<FrbBacktracedException>()
  const factory CustomNestedError1.customNested1(
    String field0,
  ) = CustomNested1;
  @Implements<FrbBacktracedException>()
  const factory CustomNestedError1.errorNested(
    CustomNestedError2 field0,
  ) = ErrorNested;
}
```

> Have a look at freezed doc - it is annotated on const factory CustomError.error0({})

where?

Reply:
With

```dart
@freezed
class CustomNestedError2 with _$CustomNestedError2 {
  @Implements<FrbBacktracedException>()
  const factory CustomNestedError2.customNested2(
    String field0,
  ) = CustomNested2;
  @Implements<FrbBacktracedException>()
  const factory CustomNestedError2.customNested2Number(
    int field0,
  ) = CustomNested2Number;
}
```

I got in the generated stuff: 

`class _$CustomNested2 implements CustomNested2 {`

error:

```
Missing concrete implementation of 'getter FrbBacktracedException.backtrace'.
Try implementing the missing method, or make the class abstract.
```

Reply:
(see my reply below)